### ISSUE TITLE
[FRONTEND][ONNX] 'Split' operator not implemented

### ISSUE THREAD

**Issue by Hkathuria** [dmlc/tvm#2513]
Getting Error while compiling onnx model as in the tutorials https://docs.tvm.ai/tutorials/nnvm/from_onnx.html

Traceback (most recent call last):
  File "sface_from_onnx.py", line 12, in <module>
    sym, params = nnvm.frontend.from_onnx(onnx_model)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/onnx.py", line 967, in from_onnx
    sym, params = g.from_onnx(graph, opset)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/onnx.py", line 822, in from_onnx
    op = self._convert_operator(op_name, inputs, attr, opset)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/onnx.py", line 923, in _convert_operator
    sym = convert_map[op_name](inputs, attrs, self._params)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/onnx.py", line 453, in _impl_v1
    for i in attr['split'][:-1]:
KeyError: 'split'

Basically 'Split' operator (as specified in onnx operator List https://github.com/onnx/onnx/blob/master/docs/Operators.md#Split ) Split a tensor into a list of tensors, along the specified 'axis'. Lengths of the parts can be specified using argument 'split'. Otherwise, the tensor is split to equal sized parts.

**Comment by Hkathuria** [comment]
Basically 'Split' operator (as specified in onnx operator List https://github.com/onnx/onnx/blob/master/docs/Operators.md#Split ) is to infer the shape of input tensor as 1D int64 output tensor .
Is there any alternative way to output the shape in 1D int64 tensor in onnx frontend ?

**Comment by Hkathuria** [comment]
@tqchen @zhreshold could you please help on this

**Comment by zhreshold** [comment]
currently topi.split takes `indices_or_sections` as splitting argument. 

@tqchen  I need clarification if this operator is fully compatible with numpy: https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.split.html or not.
More specifically, what is the behavior if integer N is used to split the array, but cannot be evenly divided? Numpy says `If indices_or_sections is an integer, N, the array will be divided into N equal arrays along axis. If such a split is not possible, an error is raised.` However, we can use `np.array_split` in such cases.

**Comment by tqchen** [comment]
I have not yet checked the unequal split when n is given. Can you run a quick exp to confirm?

**Comment by Hkathuria** [comment]
Update :  I have changed the code split Class  of below python File https://github.com/dmlc/tvm/blob/master/nnvm/python/nnvm/frontend/onnx.py
Original 
class Split(OnnxOpConverter):
    """ Operator converter for Split.
    """

    @classmethod
    def _impl_v1(cls, inputs, attr, params):
        attr['indices_or_sections'] = []
        index = 0
        for i in attr['split'][:-1]:
            index += i
            attr['indices_or_sections'].append(index)
        return AttrCvt(
            op_name='split',
            ignores=['split'])(inputs, attr, params)

Edited:
class Split(OnnxOpConverter):
    """ Operator converter for Split.
    """

    @classmethod
    def _impl_v1(cls, inputs, attr, params):
        attr['indices_or_sections'] = []
        index = 0
        for i in attr['indices_or_sections'][:-1]:
            index += i
            attr['indices_or_sections'].append(index)
        return AttrCvt(
            op_name='split',
            ignores=['split'])(inputs, attr, params)

Changes Made 
for i in attr['indices_or_sections'][:-1]:
            index += i
rather than 
for i in attr['split'][:-1]:
            index += i

I am getting the new error as  
Traceback (most recent call last):
  File "sface_from_onnx.py", line 12, in <module>
    sym, params = nnvm.frontend.from_onnx(onnx_model)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/onnx.py", line 968, in from_onnx
    sym, params = g.from_onnx(graph, opset)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/frontend/onnx.py", line 827, in from_onnx
    len(node_output), len(op.list_output_names()), op_name))
AssertionError: Number of output mismatch 32 vs 1 in Split.

Split operation is suppose to 32 outputs as per the model but is an error here.

@zhreshold Please let me know if the change is correct or something else is required. 

Thanks in advance

**Comment by zhreshold** [comment]
@Hkathuria Without manullay setting indices, will a single interger work?

```python
@classmethod
def _impl_v1(cls, inputs, attr, params):
    new_attr = {'indices_or_sections': attr['split']}
    return AttrCvt(
        op_name='split',
        ignores=['split'])(inputs, new_attr, params)
```

**Comment by Hkathuria** [comment]
@zhreshold 

@classmethod
def _impl_v1(cls, inputs, attr, params):
    new_attr = {'indices_or_sections': attr['split']}
    return AttrCvt(
        op_name='split',
        ignores=['split'])(inputs, new_attr, params)

Making the above changes I am again getting the below error:


Traceback (most recent call last):
  File "sface_from_onnx.py", line 12, in <module>
    sym, params = nnvm.frontend.from_onnx(onnx_model)
  File "/home/ubuntu/tvm_opencl/tvm/nnvm/python/nnvm/frontend/onnx.py", line 977, in from_onnx
    sym, params = g.from_onnx(graph, opset)
  File "/home/ubuntu/tvm_opencl/tvm/nnvm/python/nnvm/frontend/onnx.py", line 832, in from_onnx
    op = self._convert_operator(op_name, inputs, attr, opset)
  File "/home/ubuntu/tvm_opencl/tvm/nnvm/python/nnvm/frontend/onnx.py", line 933, in _convert_operator
    sym = convert_map[op_name](inputs, attrs, self._params)
  File "/home/ubuntu/tvm_opencl/tvm/nnvm/python/nnvm/frontend/onnx.py", line 451, in _impl_v1
    new_attr = {'indices_or_sections': attr['split']}
KeyError: 'split'

**Comment by zhreshold** [comment]
@Hkathuria  According to ONNX doc, it should have a `split` attribute: https://github.com/onnx/onnx/blob/master/docs/Operators.md#split

Can you debug the attributes of the original node in ONNX?

**Comment by tqchen** [comment]
https://github.com/dmlc/tvm/pull/2038
